### ISSUE TITLE
6.0 Deprecation List

### ISSUE THREAD

**Issue by RyanCavanaugh** [microsoft/TypeScript#54500]
We're planning the following set of deprecations for TypeScript 6.0. While TypeScript 6.0 will follow the same "soft deprecation" strategy as TypeScript 5.0, these options will be removed entirely in TypeScript 7 (the native TypeScript port).

Please let us know if these would be unreasonably disruptive for you by leaving a comment here.

---

In the two years since TypeScript 5.0, we've seen ongoing shifts in the JavaScript ecosystem:
 * Virtually every runtime environment is now "evergreen". True legacy environments (ES5) are vanishingly rare
 * Bundlers and ESM are the most common module environments, but CommonJS is still a major target. AMD and other in-browser userland module systems are much rarer than in 2012
 * Almost all packages ship as some form of module. UMD packages still exist, but virtually no new code is only available as a global variable
 * `tsconfig.json` is nearly universal as a config mechanism
 * Appetite for "stricter" typing continues to grow
 * TypeScript build performance is top of mind. Despite the gains of TypeScript 7, we still must always have performance as a key goal, and options which can't be supported in a performant way need to be more strongly justified

### Newer Default Values

 * The new default `target` is `es<current year>` - done in `#63067`
 * The new default `module` is `esnext`
 * `strict` is now `true` by default
 * `noUncheckedSideEffectImports` is now `true` by default - done in `#62443`
 * `libReplacement` is now `false` by default done in `#62391`

### `--outFile` removed

Removed for implementation simplicity's sake.
External bundlers or transpilers do this job faster, better, and more configurably.
Done in `#62981`

### Removed Flag Values

* `module: amd`, `module: umd`, and `module: systemjs` done in `#62669`
* `moduleResolution: classic` done in `#62669` 
* `esModuleInterop: false` done in `#62567`
* `allowSyntheticDefaultImports: false` done in `#62567`
* `target: es5` done in `#63067`

This also implies dropped support for the `amd-module` directive, which will no longer do anything.

### `--downlevelIteration` removed

This flag only has effects on ES5 emit, which is deprecated; see `#63071`

### Legacy `module` syntax

Early versions of TypeScript used this construct
```ts
module Foo {

}
```
which was later aliased to the modern preferred form
```ts
namespace Foo {

}
```

`module` blocks are a potential ECMAScript proposal which would conflict with this syntax, so we're fully deprecating using `module` where `namespace` is expected.

The ambient module declaration form
```ts
declare module "some-module" {
```
remains fully supported.

Done in `#62876`

See #61450 for RWC results

### DOM tweaks

`"lib": ["dom"]` will now automatically include `dom.iterable` as well

Done in `#62037`

### `libReplacement` defaults to false

This flag incurs a large number of failed module resolutions, which in turn increases the number of locations we need to watch under `--watch` and LSP scenarios. In a new project, `libReplacement` never does anything until other explicit configuration takes place, so it makes sense to turn this off by default.

Done in `#62391`

### `rootDir` default

Currently `rootDir`, if not specified, is inferred based on the set of root files found (unless `composite` is set).
In 6.0, the default `rootDir` will always be the directory containing the `tsconfig.json` file.
If a non-config-based build occurs, `rootDir` will be inferred as before.

This lets us go faster because we don't need to compute the set of input files in order to determine the structure of the output directory.

Done in `#62418`

### `types` default

Currently, the `types` value defaults to "enumerate everything in `node_modules/@types`". This is potentially *very* expensive as a normal repo setup these days might transitively pull in hundreds of `@types` packages. Modern projects almost always need only `@types/node`, `@types/jest`, or a handful of other common global-affecting packages.

In 6.0, the default `types` value will be `[]`.
*This will affect many projects*; you will likely need to add `"types": ["node"]` or a few others.

This will prevent projects from unintentionally pulling in hundreds or even thousands of unneeded declaration files at build time. Many projects we've looked at have improved their build time anywhere from 20-50% just by setting `types` appropriately.

### `alwaysStrict`

The `alwaysStrict` flag refers to inference of the `"use strict";` directive. In TypeScript 6.0, all code will be assumed to be in "strict mode", which is a [set of JS semantics](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) that mostly affects syntactic corner cases around reserved words.

This lets us be faster because it's no longer necessary to look ahead or repase on constructs like
```ts
var x = await
```
which is legal "sloppy mode" referring to a variable named `await`. This also reduces the number of flags which possibly prevent sharing of source files between projects with different compiler settings.

See `#61888` for RWC run; we found one codebase that used `static` as a parameter name.

### `no-default-lib` directives

No one really understands what this does or uses it correctly. Done in `#62435`

### Specifying commandline files when `tsconfig.json` exists is an error

Currently if you run `tsc foo.ts` in a folder where a `tsconfig.json` exists, the `tsconfig` file is completely ignored. This is very confusing. The *intended* meaning of this is also unclear; it could be that you want to use the `compilerOptions` of the config but not the `files` and/or `include` and/or `exclude`.

For clarity, in 6.0, if you run `tsc` in a directory with a `tsconfig.json`, an error will be issued instead.

> `tsconfig.json` is present but will not be loaded if files are specified on the commandline. Use `--ignore-config` to skip this error

See also #27379. Done in `#62477`

### JS Modernization

The `asserts` keyword was added to the JS language and then removed; in 6.0 this will be a parse error. See also #58453

Conditional imports/exports fallback lookups are deprecated; see #50762
If your name isn't Andrew Branch and you know what this even is about, please let us know.

See also #51000 / #51424 / #51909 / #54500

**Comment by RyanCavanaugh** [comment]
This is a tracking issue, so I'm locking this since I don't want a flood of "You should deprecate X" comments here. Post on the relevant thread for X instead, and if it seems like a possible fit for deprecation, we'll cross-link here.

**Comment by andrewbranch** [comment]
Module stuff:

- Deprecate [conditional imports/exports fallback lookups](https://github.com/microsoft/TypeScript/issues/50762)
- Deprecate `--moduleResolution classic`
  - Change default `moduleResolution` for `--module es2015` et al.
- Deprecate `--moduleResolution node`/`node10`
  - Change default `moduleResolution` for `--module commonjs`
- Deprecate `allowSyntheticDefaultExports` and `esModuleInterop` (make the latter always true)

**Comment by DanielRosenwasser** [comment]
We've talked quite a bit about moving `strictNullChecks` and `noImplicitAny` to being on by default. In order to transition our userbase, would it make sense to require them to either

* set to `true` either explicitly or implicitly (i.e. via `--strict`) or
* explicitly be set to `false`?

**Comment by RyanCavanaugh** [comment]
`tsc foo.ts` not applying `tsconfig.json` - See #57159 and *many* others

**Comment by RyanCavanaugh** [comment]
`no-default-lib`

**Comment by RyanCavanaugh** [comment]
`amd-module`

**Comment by RyanCavanaugh** [comment]
Mix `tsconfig.json` with extra files #27379

**Comment by sheetalkamat** [comment]
May be default `types` should be `[]` verses `undefined`

**Comment by RyanCavanaugh** [comment]
#58453 error on `asserts` in `import` statements

**Comment by RyanCavanaugh** [comment]
Multi-file global value namespaces

**Comment by andrewbranch** [comment]
`--rootDirs`? https://www.typescriptlang.org/tsconfig/#rootDirs

**Comment by andrewbranch** [comment]
Import assertions should start erroring (saying to use import attributes instead) in `--module esnext` and `--module preserve`. They will already error in `--module nodenext` when #60534 is done.

**Comment by andrewbranch** [comment]
I forgot to include `--baseUrl` earlier

**Comment by RyanCavanaugh** [comment]
`module Foo {` (synonym of `namespace Foo {`), see #61450 for RWC results

**Comment by rbuckton** [comment]
`enum` declaration merging. See #61262 for RWC results.

**Comment by jakebailey** [comment]
`DOM` without `DOM.Iterable`, ES5

**Comment by RyanCavanaugh** [comment]
Force `alwaysStrict` to `true` ?

**Comment by jakebailey** [comment]
`preserveConstEnums=false`?

As in, always emit the enum object unconditionally?

**Comment by mjames-c** [comment]
clarification on `allowSyntheticDefaultExports: true` removal -- the docs currently state the following for the `allowSyntheticDefaultExports` option:

> true if [esModuleInterop](https://www.typescriptlang.org/tsconfig/#esModuleInterop) is enabled, [module](https://www.typescriptlang.org/tsconfig/#module) is system, or [moduleResolution](https://www.typescriptlang.org/tsconfig/#module-resolution) is bundler; false otherwise.

will TSv7 continue to adhere to the above? if so, doesn't that actually imply the reverse -- that `allowSyntheticDefaultExports: true` will always be set (since `esModuleInterop` will be forced `true` in TSv7)

**Comment by mjames-c** [comment]
I have a similar question for removal of `esModuleInterop: false` -- the docs have this:

> true if [module](https://www.typescriptlang.org/tsconfig/#module) is node16, nodenext, or preserve; false otherwise.

in TSv7 will `esModuleInterop` continue to adhere to this logic or will it always be `true` regardless of the `module` type the user selects? I'm also curious if users will always want `esModuleInterop: true` as it seems to increase the size of compiled code/bundles (although perhaps that's negligible)

**Comment by JustFly1984** [comment]
There is an issue with supporting ts/tsx extensions. In VSCode it doesn't see the difference between ts and tsx in imports, but in runtime it does. E.g. - if AI autoimports something with incorrect extension, you don't see it until build or even in runtime.
Especally bad in compiling scenarios like react-native, wasting so much time. Can't be detected with tsc, but can be detected with eslint/oxlint/biomejs.

**Comment by jakebailey** [comment]
@JustFly1984 That has nothing to do with this thread, please file a new issue

**Comment by andrewbranch** [comment]
@mjames-c no, the behavior in all modes will be `esModuleInterop: true`. The behavior has long been _the_ standard way to transpile ESM imports to CommonJS and giving people the option to do something non-standard isn’t helping anyone. The way to escape its impact on emit is to migrate away from compiling to CommonJS.

**Comment by majames** [comment]
@andrewbranch ok cool. what about `allowSyntheticDefaultExports`?

**Comment by andrewbranch** [comment]
The type checking behavior of `esModuleInterop: true` completely subsumes that of `allowSyntheticDefaultExports`, so it will also be removed, as toggling it wouldn’t do anything.

**Comment by jakebailey** [comment]
I swear we also mentioned `compileOnSave`, but I don't see that listed.

**Comment by mjames-c** [comment]
> Removed Flag Values
> 
> `module: amd`, `module: umd`, and `module: systemjs`

does this mean that `module: commonjs` will be brought forward?

**Comment by fregante** [comment]
CommonJS itself is not going anywhere, Node has officially no intention to possibly ever deprecate it.

**Comment by mjames-c** [comment]
> CommonJS itself is not going anywhere, Node has officially no intention to possibly ever deprecate it.

ok makes sense, i assume the `moduleResolution` default must be changing to `node16` or similar then? (since `node10` is going away and `nodenext` isn't allowed: `error TS5110: Option 'module' must be set to 'NodeNext' when option 'moduleResolution' is set to 'NodeNext'.`)

**Comment by mjames-c** [comment]
actually i'm not sure this currently makes sense given the current docs -- what will be the default resolution strategy for `"module": "commonjs"` when `node10` module resolution is removed? the docs currently enforce `module: commonjs` can't be used with the more up to date module resolution strategies:

<img width="888" height="82" alt="Image" src="https://github.com/user-attachments/assets/49ec463e-ec24-427c-a9e8-3dc300f52d9a" />

https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext

**Comment by andrewbranch** [comment]
`--module commonjs` will use `--moduleResolution bundler`, which is currently a prohibited combination in TS 5.X, but will be flipped on in 6.0

**Comment by darcyrush** [comment]
@andrewbranch just confirming your comment [in this thread](https://github.com/microsoft/TypeScript/issues/60705) still stands since I see no explicit mention here; that `--module node16` will be deprecated in favour of `node18`?

**Comment by darcyrush** [comment]
Also curious on the status of `experimentalDecorators` (and `emitDecoratorMetadata`) in Corsa now that ECMAScript decorators are supported in Strada

**Comment by lishaduck** [comment]
> If your name isn't Andrew Branch and you know what this even is about, please let us know.

Having read the [attw docs](https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FallbackCondition.md) & this issue too many times, I actually do (now) :P
(you asked for it lol)

**Comment by DanielRosenwasser** [comment]
> I swear we also mentioned `compileOnSave`, but I don't see that listed.

Chatted with @jakebailey - we won't be bringing that into TS 7.0, but `compileOnSave` is a top-level setting, not in `compilerOptions`. That means we wouldn't deprecate it in 6.0, we would just not have it work in 7.0.

**Comment by DanielRosenwasser** [comment]
> Also curious on the status of `experimentalDecorators` (and `emitDecoratorMetadata`) in Corsa now that ECMAScript decorators are supported in Strada

I think the usage is sufficiently high for both to the point that we likely won't deprecate either. For example, see the download count of [reflect-metadata](https://www.npmjs.com/package/reflect-metadata) as a proxy for usage.

Deprecating `emitDecoratorMetadata` comes up quite a bit because it requires the same architectural problems we want to avoid in our emitter (type-driven emit), and because it doesn't align with standards at the moment. But given the usage of both, I think it is going to still be in 7.0.

**Comment by silverwind** [comment]
> "lib": ["dom"] will now automatically include dom.iterable as well

Will this also include `dom.asynciterable`?

**Comment by jakebailey** [comment]
asynciterable requires ES2018, so I would think not, no. See also #62196

**Comment by jakebailey** [comment]
If we do something like #62025 instead of "proper" `references`, then we could force the inclusion of `DOM.AsyncIterable` when `ES2018` is loaded, but that solution seemed hacky.

**Comment by miscellaneo** [comment]
> > Also curious on the status of `experimentalDecorators` (and `emitDecoratorMetadata`) in Corsa now that ECMAScript decorators are supported in Strada
> 
> I think the usage is sufficiently high for both to the point that we likely won't deprecate either. For example, see the download count of [reflect-metadata](https://www.npmjs.com/package/reflect-metadata) as a proxy for usage.
> 
> Deprecating `emitDecoratorMetadata` comes up quite a bit because it requires the same architectural problems we want to avoid in our emitter (type-driven emit), and because it doesn't align with standards at the moment. But given the usage of both, I think it is going to still be in 7.0.

I doubt this is an insight you haven't already heard from more notable people, but I want to highlight that 6.0 is likely the last chance to steer people away from `experimentalDecorators`.

If you deprecate it here and give people a big warning saying "stop using this", they'll still have a couple of years before 7.0.  There can be further big shifts during those years; maybe there'll be enough outcry to walk back to deprecation, maybe the deprecation will force the standard to update, maybe something else will happen and the whole thing is moot.  Things can happen.

But if you keep it on for 6.0, you give up that window to fit it in with the other deprecations.  People will keep using it, and they'll rightfully expect it in 7.0.  You can't take it out after that, because this was the big event, the big rewrite where you have room to make huge changes like this.  You can't justify taking it out outside of here.  And in 2027, 2029, 2031, however long people keep using TypeScript, two separate kinds of decorators will need to be supported.  People will still need to learn the difference between two things with the same name.  People will continue to run into the edge cases that come with a non-standard feature.  Barring a big change in decorator specification, there will be miscommunication and confusion on the topic, forever.

It won't be easy to sunset `experimentalDecorators`.  It's useful and it's been around a while, so it can't be easy.  But if it's a thing worth doing, I implore you to take this opportunity to do so.

**Comment by GabenGar** [comment]
These all sound like reasons for keeping status quo as is. Microsoft affecting the spec indirectly using its consumers as cannon fodder is one of the better outcomes for it. It's entirely microsoft induced problem which requires microsoft solution, so by the year 2027 you will just ask copilot to resolve the edge cases instead.

**Comment by andrewbranch** [comment]
> [@andrewbranch](https://github.com/andrewbranch) just confirming your comment [in this thread](https://github.com/microsoft/TypeScript/issues/60705) still stands since I see no explicit mention here; that `--module node16` will be deprecated in favour of `node18`?

Yes, that's still the plan; sorry we missed that in the summary writeup.

**Comment by connorshea** [comment]
Regarding the `types: []` change, is there a good way to auto-detect what was actually used as part of typechecking (node, jest, etc.), as a means of providing an automatic migration path once 6.0 releases?

**Comment by jakebailey** [comment]
Practically speaking, no, there's no way to detect if a certain type "matters" except to speculatively not load it and see if things break.

**Comment by connorshea** [comment]
I think this thread (and/or #62508, I don't know if that thread is intended for random comments like this) are missing the change from #62443?

**Comment by connorshea** [comment]
the `libReplacement: false` default is also missing in this issue I think? #62391

**Comment by connorshea** [comment]
Also this issue says the new default target is `es2015`, but that's going to be the lowest allowed value according to #62198, I assume that's meant to say the new default target is `esnext`?

(As an aside, maybe it should be considered to go so far as to deprecate es2015 and es2016 as well, to push things forward a bit? Maybe introduce a rolling deprecation schedule for older es year options?)

**Comment by fregante** [comment]
Should new projects default to a 11-year-old spec? I think a reasonable _default_ would be 3-4 years; **the minority who needs broader compatibility can still configure it.**

With all browsers being evergreen, supporting browsers/engines that haven't been updated since 2022 already means supporting unpatched and unsafe systems. 

ES2022 for example is supported by Node 18, which is two LTS versions behind.

**Comment by RyanCavanaugh** [comment]
> the libReplacement: false default is also missing in this issue

Thanks! Fixed

> Should new projects default to a 11-year-old spec?

Absolutely not, and that's why `tsc --init` specifies `target: esnext`

The question is what should happen to *existing* projects that never specified a target. What should happen to them? The problem is that our downlevelings have never been perfect, and in some cases have been *very* imperfect, particular when it comes to class field initialization order. In particular, silently changing a project from `useDefineForClassFields: false` to `true` implies a very real possibility of breaking these programs at runtime, often in ways that are subtle and hard to debug. That's a very large downside that we don't want to incur especially as 6.0/7.0 is incurring enough other breaks at the same time.

**Comment by andrewbranch** [comment]
Edited the OP to reflect `false` is the value being deprecated for both `esModuleInterop` and `allowSyntheticDefaultImports` (#62567). It previously said `allowSyntheticDefaultImports: true` was being deprecated.

**Comment by jakebailey** [comment]
Since ES5 will be deprecated, we should also deprecate `downlevelIteration`.

**Comment by jakebailey** [comment]
I'm silly, that's clearly in #62196. Yay collapsed github threads

**Comment by fry69** [comment]
> allowSyntheticDefaultExports: false

Please correct this to:

```
allowSyntheticDefaultImports: false,
```

LLMs (GPT-5.1 Thinking) picked this up and hallucinated on this. Chat interaction below for illustration what brought me here:

---

You’re absolutely right to be suspicious there — and nice catch.

### 1. Does `allowSyntheticDefaultExports` exist?

**No.** There is **no** TypeScript compiler option called `allowSyntheticDefaultExports`.

The real option is:

```jsonc
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true
  }
}
```

You can see it in the official TSConfig reference and CLI options list as `allowSyntheticDefaultImports` (no “Exports” version anywhere).

So if `allowSyntheticDefaultExports` “doesn’t work” for you, that’s because it simply isn’t a recognized option.

### 2. Why did “allowSyntheticDefaultExports” show up anywhere?

The confusion comes from the **TypeScript 6.0 deprecation candidates** issue on GitHub, which has this line under “Removed Flag Values”:

> `esModuleInterop: false`
> `allowSyntheticDefaultExports: false`

That **looks** like a compiler option name, but it does not match the real option (`allowSyntheticDefaultImports`). It’s almost certainly a typo in that issue:

* The actual option is `allowSyntheticDefaultImports`.
* `esModuleInterop` already **enables** `allowSyntheticDefaultImports` automatically.

What they’re trying to say is:

> In TS 6, you won’t be allowed to set **`esModuleInterop: false`** and **`allowSyntheticDefaultImports: false`** anymore — modern ESM/CJS interop becomes the standard behavior.

But the name they wrote (“Exports”) is just wrong.

### 3. What does `allowSyntheticDefaultImports` actually do?

Quick refresher:

* `allowSyntheticDefaultImports` lets you write:

  ```ts
  import React from "react";
  ```

  even when the module only has CommonJS-style exports and no real `export default`.

* It **only affects type-checking**, not emitted JS. `esModuleInterop` is the one that adjusts the emit / helper functions while also turning this on.

In modern TS projects, the common advice is:

* Just set `"esModuleInterop": true` (which also turns on `allowSyntheticDefaultImports`),
* And **don’t** bother explicitly toggling `allowSyntheticDefaultImports` unless a tool (like an IDE plugin) reads that flag directly.

### 4. What to change in your mental model / config

* Use **only**:

  ```jsonc
  {
    "compilerOptions": {
      "esModuleInterop": true
      // optional: "allowSyntheticDefaultImports": true
    }
  }
  ```

* Ignore any mention of `allowSyntheticDefaultExports` — it’s not a real flag, and if you see it in docs/notes, treat it as a typo.

Also: in my earlier explanation + example I repeated that typo from the 6.0 issue — that’s on me. The correct “removed flag values” for TS 6 are:

* `esModuleInterop: false`
* `allowSyntheticDefaultImports: false` (not “…Exports”)

Your instinct here was spot on.

**Comment by michael-small** [comment]
> In particular, silently changing a project from `useDefineForClassFields: false` to `true` implies a very real possibility of breaking these programs at runtime, often in ways that are subtle and hard to debug. That's a very large downside that we don't want to incur especially as 6.0/7.0 is incurring enough other breaks at the same time.

Speaking of `useDefineForClassFields`, what is the scope of that in v6/v7? Maybe that is implied by that discussion about the targets but tbh I'm a bit dense.

Until the big new blog post I have referred back to the proposed plan in https://github.com/microsoft/TypeScript/issues/60284 for a 6.0 deprecation and the question of the "idea that we just flip the emit and errors over in 6.0 and let people set `useDefineForClassFields`+`ignoreDeprecations` until 6.5". 

Speaking from a project that does specify a target. Didn't see `useDefineForClassFields` in the 6.0 deprecations PRs and been working on this 60284's proposed guidelines as a "hard deadline" to strict-ify a big legacy codebase due to uncaught errors when non-strict projects remove `useDefineForClassFields`.

**Comment by RyanCavanaugh** [comment]
> Speaking of useDefineForClassFields, what is the scope of that in v6/v7?

This will remain supported. While it's the sort of thing we'd love to remove if we could, a lot of very large codebases depend on it, and in cases where you actually need its behavior, it's extremely difficult to migrate away from it.

**Comment by danilofuchs** [comment]
Will `noUncheckedIndexAccess` turned on by default in strict mode?

**Comment by RyanCavanaugh** [comment]
> Will `noUncheckedIndexAccess` turned on by default in strict mode?

No

**Comment by RyanCavanaugh** [comment]
Edited to add target: es5 and downlevelIteration (entirely) being deprecated.

**Comment by RyanCavanaugh** [comment]
Edited to un-deprecate `enum` merging and cross-namespace value referencing; these are in 7.0 already and removing them wouldn't produce the upside I had hoped for.

**Comment by DanielRosenwasser** [comment]
I think we also passed on conditional import fallbacks: https://github.com/microsoft/TypeScript/issues/50762

**Comment by HolgerJeromin** [comment]
Wow. You are removing `module:none` !!!

Every public place is writing about deprecating some "old" module options:
* `module: amd, module: umd, and module: systemjs` (this issue text)
*  "The following flag values are no longer supported --module amd, --module umd, --module systemjs" [6.0 beta blog post](https://devblogs.microsoft.com/typescript/announcing-typescript-6-0-beta/#deprecated:-amd-umd-and-systemjs-values-of-module)

But in the middle of nowhere you find the text "We are deprecating [module=none] because its behavior is inconsistent and weird in Strada" in a comment in the PR https://github.com/microsoft/TypeScript/pull/62669#issuecomment-3445053364

And using ts6.0-beta this is real: 
> .../tsconfig.json(5,19): error TS5107: Option 'module=None' is deprecated and will stop functioning in TypeScript 7.0. Specify compilerOption '"ignoreDeprecations": "6.0"' to silence this error

I do not know what other developers think, but keeping only module: es20xx and commonJs is very challenging.
IMO this will keep many projects stuck with ts6.0 for years.

**Comment by n9** [comment]
> I do not know what other developers think, but keeping only module: es20xx and commonJs is very challenging.

I agree. The `amd` module format (with `outFile`) is useful for enterprise systems where TypeScript serves as the foundation for internal DSLs.

**Comment by GabenGar** [comment]
Did you just throw `enterprise` in your sentence so it would look more legit? Expecting a dependency to be forward-compatible across several semver major releases in regards to legacy behaviour is quite a business strategy. If it's your company specifically, tell your boss it's time to allocate more budget in order to clean up some tech debt.
